server/auth/auth.go

```diff
@@ -190,6 +200,113 @@ type AuthManager struct {
 
 	pendingMessagesMtx sync.Mutex
 	pendingMessages    map[account.AccountID]map[uint64]*timedMessage
+
+	violationMtx    sync.Mutex
+	violationCounts map[account.AccountID]violationCounter
+	matchOutcomes   map[account.AccountID]*latest
+	preimgOutcomes  map[account.AccountID]*latest
+}
+
+// violation badness
+const (
+	// preimage miss
+	preimageMissScore = 2 // book spoof, no match, no stuck funds
+
+	// failure to act violations
+	noSwapAsMakerScore   = 4  // book spoof, match with taker order affected, no stuck funds
+	noSwapAsTakerScore   = 11 // maker has contract stuck for 20 hrs
+	noRedeemAsMakerScore = 7  // taker has contract stuck for 8 hrs
+	noRedeemAsTakerScore = 1  // just dumb, counterparty not inconvenienced
+
+	successScore = -1 // offsets the violations
+
+	defaultBanScore = 20
```

Comment:
Just want to mention here that from a UI perspective, I would much prefer to have a score where positive is good and negative is bad. I like the natural semantics of a positive = good system, where you don't have a negatively-connotated "ban score", you have some kind of "trader score", and the user can take comfort in seeing their score is at max value. Not saying it has to be that way here. Since it's not possible to be on [0, 100] in this scoring system, I'd want to apply some kind of transformation somewhere in the pipeline before display anyway.